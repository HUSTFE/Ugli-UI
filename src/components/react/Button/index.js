import React from 'react'
import PropTypes from 'prop-types'

import styles from '@style/Button/index.sass'
import { warning } from '@shared/warning'
import { computeSize } from '@shared/size'

/*
const defaultTheme = {
  startColor: '#fbc8d4',
  endColor: '#d8a3fd',
}
TODO: set a shared default theme somewhere else
There IS a default theme, but should not be applied here,
the only thing sub-component
*/
// So, the priority is like: props > context > default, which means we can't just set default props simply
// kind of disgusting...
const Button = (props, context) => {
  const { ctxStartColor = '#fbc8d4', ctxEndColor = '#d8a3fd' } = context
  // TODO: warning(!useTheme, 'To get the best experience of Ugli-UI, we suggest to use component `ThemeProvider`.')
  const { startColor = ctxStartColor, endColor = ctxEndColor, text, size, style, onClick } = props
  return (
    <button
      className={styles.button}
      onClick={onClick}
      style={{
        fontSize: computeSize(size),
        background: `linear-gradient(-45deg, ${startColor} 30%, ${endColor} 70%)`,
        ...style,
      }}
    >
      <span>{text}</span>
    </button>
  )
}

Button.propTypes = {
  startColor: PropTypes.string,
  endColor: PropTypes.string,
  size: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  style: PropTypes.shape(),
  onClick: PropTypes.func,
  text: PropTypes.string.isRequired,
}
Button.defaultProps = {
  startColor: undefined,
  endColor: undefined,
  size: '16',
  style: {},
  onClick: () => warning(true, 'onClick prop should be passed to Bottom.'),
}
Button.contextTypes = {
  startColor: PropTypes.string,
  endColor: PropTypes.string,
}

export default Button
