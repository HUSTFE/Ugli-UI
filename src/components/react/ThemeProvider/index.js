import React from 'react'
import PropTypes from 'prop-types'

// TODO: In react 16.3, old context api will be deprecated,
// this component will be refactored.
export default class ThemeProvider extends React.Component {
  static propTypes = {
    startColor: PropTypes.string.isRequired,
    endColor: PropTypes.string.isRequired,
  }

  static childContextTypes = ThemeProvider.propTypes

  getChildContext() {
    return {
      startColor: this.props.startColor,
      endColor: this.props.endColor,
    }
  }

  render() {
    return this.props.children
  }

  shouldComponentUpdate(nextProps) {
    const { startColor, endColor } = nextProps
    if (startColor === this.props.startColor && endColor === this.props.endColor) {
      return false
    }
    return true
  }
}
