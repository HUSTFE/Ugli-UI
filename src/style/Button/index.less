.ugli-button {
  background-color: aqua;
}
