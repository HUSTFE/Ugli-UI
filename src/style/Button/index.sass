.button
  display: inline-block
  border: none
  text-align: center
  line-height: 1.5
  box-sizing: border-box
  transition: background-color .3s
  background: transparent
  cursor: pointer
  outline: none
  padding: .5em 1em
  user-select: none
