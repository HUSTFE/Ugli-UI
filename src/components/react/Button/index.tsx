import * as React from 'react';

import cls from '../../../style/Button/index.less';

export default class Button extends React.Component {
  render() {
    return <button className={cls['ugli-button']}>hehe</button>;
  }
}
