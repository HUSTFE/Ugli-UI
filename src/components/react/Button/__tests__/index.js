import React from 'react'
import renderer from 'react-test-renderer'
import Button from '../index'
import ThemeProvider from '../../ThemeProvider/index'

describe('Button', () => {
  it('Basic button with text', () => {
    const component = renderer.create(
      <ThemeProvider startColor="green" endColor="red">
        <Button text="Button" />
      </ThemeProvider>,
    )
    const json = component.toJSON()
    expect(json)
      .toMatchSnapshot()
  })
})
