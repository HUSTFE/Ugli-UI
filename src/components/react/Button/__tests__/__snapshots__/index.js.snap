// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Basic button with text 1`] = `
<button
  className="button"
  onClick={[Function]}
  style={
    Object {
      "background": "linear-gradient(-45deg, #fbc8d4 30%, #d8a3fd 70%)",
      "fontSize": "16px",
    }
  }
>
  <span>
    Button
  </span>
</button>
`;
